File: PolicyPlus/AdmlFile.vb
Imports System.Xml
Public Class AdmlFile
    Public SourceFile As String
    Public Revision As Decimal
    Public DisplayName As String
    Public Description As String
    Public StringTable As New Dictionary(Of String, String)
    Public PresentationTable As New Dictionary(Of String, Presentation)
    Private Sub New()
    End Sub
    Public Shared Function Load(File As String) As AdmlFile
        ' ADML documentation: https://technet.microsoft.com/en-us/library/cc772050(v=ws.10).aspx
        Dim adml As New AdmlFile
        adml.SourceFile = File
        Dim xmlDoc As New XmlDocument
        xmlDoc.Load(File)
        ' Load ADML metadata
        Dim policyDefinitionResources As XmlNode = xmlDoc.GetElementsByTagName("policyDefinitionResources")(0)
        adml.Revision = Decimal.Parse(policyDefinitionResources.Attributes("revision").Value, Globalization.CultureInfo.InvariantCulture)
        For Each child As XmlNode In policyDefinitionResources.ChildNodes
            Select Case child.LocalName
                Case "displayName"
                    adml.DisplayName = child.InnerText
                Case "description"
                    adml.Description = child.InnerText
            End Select
        Next
        ' Load localized strings
        Dim stringTableList = xmlDoc.GetElementsByTagName("stringTable")
        If stringTableList.Count > 0 Then
            Dim stringTable = stringTableList(0)
            For Each stringElement As XmlNode In stringTable.ChildNodes
                If stringElement.LocalName <> "string" Then Continue For
                Dim key = stringElement.Attributes("id").Value
                Dim value = stringElement.InnerText
                adml.StringTable.Add(key, value)
            Next
        End If
        ' Load presentations (UI arrangements)
        Dim presTableList = xmlDoc.GetElementsByTagName("presentationTable")
        If presTableList.Count > 0 Then
            Dim presTable = presTableList(0)
            For Each presElement As XmlNode In presTable.ChildNodes
                If presElement.LocalName <> "presentation" Then Continue For
                Dim presentation As New Presentation
                presentation.Name = presElement.Attributes("id").Value
                For Each uiElement As XmlNode In presElement.ChildNodes
                    Dim presPart As PresentationElement = Nothing
                    Select Case uiElement.LocalName
                        Case "text"
                            Dim textPart As New LabelPresentationElement
                            textPart.Text = uiElement.InnerText
                            presPart = textPart
                        Case "decimalTextBox"
                            Dim decTextPart As New NumericBoxPresentationElement
                            decTextPart.DefaultValue = AttributeOrDefault(uiElement, "defaultValue", 1)
                            decTextPart.HasSpinner = AttributeOrDefault(uiElement, "spin", True)
                            decTextPart.SpinnerIncrement = AttributeOrDefault(uiElement, "spinStep", 1)
                            decTextPart.Label = uiElement.InnerText
                            presPart = decTextPart
                        Case "textBox"
                            Dim textPart As New TextBoxPresentationElement
                            For Each textboxInfo As XmlNode In uiElement.ChildNodes
                                Select Case textboxInfo.LocalName
                                    Case "label"
                                        textPart.Label = textboxInfo.InnerText
                                    Case "defaultValue"
                                        textPart.DefaultValue = textboxInfo.InnerText
                                End Select
                            Next
                            presPart = textPart
                        Case "checkBox"
                            Dim checkPart As New CheckBoxPresentationElement
                            checkPart.DefaultState = AttributeOrDefault(uiElement, "defaultChecked", False)
                            checkPart.Text = uiElement.InnerText
                            presPart = checkPart
                        Case "comboBox"
                            Dim comboPart As New ComboBoxPresentationElement
                            comboPart.NoSort = AttributeOrDefault(uiElement, "noSort", False)
                            For Each comboInfo As XmlNode In uiElement.ChildNodes
                                Select Case comboInfo.LocalName
                                    Case "label"
                                        comboPart.Label = comboInfo.InnerText
                                    Case "default"
                                        comboPart.DefaultText = comboInfo.InnerText
                                    Case "suggestion"
                                        comboPart.Suggestions.Add(comboInfo.InnerText)
                                End Select
                            Next
                            presPart = comboPart
                        Case "dropdownList"
                            Dim dropPart As New DropDownPresentationElement
                            dropPart.NoSort = AttributeOrDefault(uiElement, "noSort", False)
                            dropPart.DefaultItemID = AttributeOrNull(uiElement, "defaultItem")
                            dropPart.Label = uiElement.InnerText
                            presPart = dropPart
                        Case "listBox"
                            Dim listPart As New ListPresentationElement
                            listPart.Label = uiElement.InnerText
                            presPart = listPart
                        Case "multiTextBox"
                            Dim multiTextPart As New MultiTextPresentationElement
                            multiTextPart.Label = uiElement.InnerText
                            presPart = multiTextPart
                    End Select
                    If presPart IsNot Nothing Then
                        If uiElement.Attributes("refId") IsNot Nothing Then presPart.ID = uiElement.Attributes("refId").Value
                        presPart.ElementType = uiElement.LocalName
                        presentation.Elements.Add(presPart)
                    End If
                Next
                adml.PresentationTable.Add(presentation.Name, presentation)
            Next
        End If
        Return adml
    End Function
End Class

File: PolicyPlus/AdmlFile.cs
using System.Collections.Generic;
using System.Xml;
using Microsoft.VisualBasic.CompilerServices;

namespace PolicyPlus
{
    public class AdmlFile
    {
        public string SourceFile;
        public decimal Revision;
        public string DisplayName;
        public string Description;
        public Dictionary<string, string> StringTable = new Dictionary<string, string>();
        public Dictionary<string, Presentation> PresentationTable = new Dictionary<string, Presentation>();
        private AdmlFile()
        {
        }
        public static AdmlFile Load(string File)
        {
            // ADML documentation: https://technet.microsoft.com/en-us/library/cc772050(v=ws.10).aspx
            var adml = new AdmlFile();
            adml.SourceFile = File;
            var xmlDoc = new XmlDocument();
            xmlDoc.Load(File);
            // Load ADML metadata
            var policyDefinitionResources = xmlDoc.GetElementsByTagName("policyDefinitionResources")[0];
            adml.Revision = decimal.Parse(policyDefinitionResources.Attributes["revision"].Value, System.Globalization.CultureInfo.InvariantCulture);
            foreach (XmlNode child in policyDefinitionResources.ChildNodes)
            {
                switch (child.LocalName ?? "")
                {
                    case "displayName":
                        {
                            adml.DisplayName = child.InnerText;
                            break;
                        }
                    case "description":
                        {
                            adml.Description = child.InnerText;
                            break;
                        }
                }
            }
            // Load localized strings
            var stringTableList = xmlDoc.GetElementsByTagName("stringTable");
            if (stringTableList.Count > 0)
            {
                var stringTable = stringTableList[0];
                foreach (XmlNode stringElement in stringTable.ChildNodes)
                {
                    if (stringElement.LocalName != "string")
                        continue;
                    string key = stringElement.Attributes["id"].Value;
                    string value = stringElement.InnerText;
                    adml.StringTable.Add(key, value);
                }
            }
            // Load presentations (UI arrangements)
            var presTableList = xmlDoc.GetElementsByTagName("presentationTable");
            if (presTableList.Count > 0)
            {
                var presTable = presTableList[0];
                foreach (XmlNode presElement in presTable.ChildNodes)
                {
                    if (presElement.LocalName != "presentation")
                        continue;
                    var presentation = new Presentation();
                    presentation.Name = presElement.Attributes["id"].Value;
                    foreach (XmlNode uiElement in presElement.ChildNodes)
                    {
                        PresentationElement presPart = null;
                        switch (uiElement.LocalName ?? "")
                        {
                            case "text":
                                {
                                    var textPart = new LabelPresentationElement();
                                    textPart.Text = uiElement.InnerText;
                                    presPart = textPart;
                                    break;
                                }
                            case "decimalTextBox":
                                {
                                    var decTextPart = new NumericBoxPresentationElement();
                                    decTextPart.DefaultValue = Conversions.ToUInteger(uiElement.AttributeOrDefault("defaultValue", 1));
                                    decTextPart.HasSpinner = Conversions.ToBoolean(uiElement.AttributeOrDefault("spin", true));
                                    decTextPart.SpinnerIncrement = Conversions.ToUInteger(uiElement.AttributeOrDefault("spinStep", 1));
                                    decTextPart.Label = uiElement.InnerText;
                                    presPart = decTextPart;
                                    break;
                                }
                            case "textBox":
                                {
                                    var textPart = new TextBoxPresentationElement();
                                    foreach (XmlNode textboxInfo in uiElement.ChildNodes)
                                    {
                                        switch (textboxInfo.LocalName ?? "")
                                        {
                                            case "label":
                                                {
                                                    textPart.Label = textboxInfo.InnerText;
                                                    break;
                                                }
                                            case "defaultValue":
                                                {
                                                    textPart.DefaultValue = textboxInfo.InnerText;
                                                    break;
                                                }
                                        }
                                    }
                                    presPart = textPart;
                                    break;
                                }
                            case "checkBox":
                                {
                                    var checkPart = new CheckBoxPresentationElement();
                                    checkPart.DefaultState = Conversions.ToBoolean(uiElement.AttributeOrDefault("defaultChecked", false));
                                    checkPart.Text = uiElement.InnerText;
                                    presPart = checkPart;
                                    break;
                                }
                            case "comboBox":
                                {
                                    var comboPart = new ComboBoxPresentationElement();
                                    comboPart.NoSort = Conversions.ToBoolean(uiElement.AttributeOrDefault("noSort", false));
                                    foreach (XmlNode comboInfo in uiElement.ChildNodes)
                                    {
                                        switch (comboInfo.LocalName ?? "")
                                        {
                                            case "label":
                                                {
                                                    comboPart.Label = comboInfo.InnerText;
                                                    break;
                                                }
                                            case "default":
                                                {
                                                    comboPart.DefaultText = comboInfo.InnerText;
                                                    break;
                                                }
                                            case "suggestion":
                                                {
                                                    comboPart.Suggestions.Add(comboInfo.InnerText);
                                                    break;
                                                }
                                        }
                                    }
                                    presPart = comboPart;
                                    break;
                                }
                            case "dropdownList":
                                {
                                    var dropPart = new DropDownPresentationElement();
                                    dropPart.NoSort = Conversions.ToBoolean(uiElement.AttributeOrDefault("noSort", false));
                                    dropPart.DefaultItemID = Conversions.ToInteger(uiElement.AttributeOrNull("defaultItem"));
                                    dropPart.Label = uiElement.InnerText;
                                    presPart = dropPart;
                                    break;
                                }
                            case "listBox":
                                {
                                    var listPart = new ListPresentationElement();
                                    listPart.Label = uiElement.InnerText;
                                    presPart = listPart;
                                    break;
                                }
                            case "multiTextBox":
                                {
                                    var multiTextPart = new MultiTextPresentationElement();
                                    multiTextPart.Label = uiElement.InnerText;
                                    presPart = multiTextPart;
                                    break;
                                }
                        }
                        if (presPart is not null)
                        {
                            if (uiElement.Attributes["refId"] is not null)
                                presPart.ID = uiElement.Attributes["refId"].Value;
                            presPart.ElementType = uiElement.LocalName;
                            presentation.Elements.Add(presPart);
                        }
                    }
                    adml.PresentationTable.Add(presentation.Name, presentation);
                }
            }
            return adml;
        }
    }
}